com.unity.ml-agents/Runtime/Agent.cs

```diff
@@ -889,27 +894,30 @@ public virtual void Heuristic(in ActionBuffers actionsOut)
             // Disable deprecation warnings so we can call the legacy overload.
 #pragma warning disable CS0618
 
+            Array.Clear(m_LegacyHeuristicCache, 0, m_LegacyHeuristicCache.Length);
             // The default implementation of Heuristic calls the
             // obsolete version for backward compatibility
             switch (m_PolicyFactory.BrainParameters.VectorActionSpaceType)
             {
                 case SpaceType.Continuous:
-                    Heuristic(actionsOut.ContinuousActions.Array);
+                    Heuristic(m_LegacyHeuristicCache);
+                    Array.Copy(m_LegacyHeuristicCache, actionsOut.ContinuousActions.Array, m_LegacyActionCache.Length);
                     actionsOut.DiscreteActions.Clear();
                     break;
                 case SpaceType.Discrete:
-                    var convertedOut = Array.ConvertAll(actionsOut.DiscreteActions.Array, x => (float)x);
-                    Heuristic(convertedOut);
+                    Heuristic(m_LegacyHeuristicCache);
                     var discreteActionSegment = actionsOut.DiscreteActions;
                     for (var i = 0; i < actionsOut.DiscreteActions.Length; i++)
                     {
-                        discreteActionSegment[i] = (int)convertedOut[i];
+                        discreteActionSegment[i] = (int)m_LegacyHeuristicCache[i];
                     }
                     actionsOut.ContinuousActions.Clear();
                     break;
             }
 #pragma warning restore CS0618
 
+            // Send heuristic buffers to actuators if they implement IHeuristic
+            m_ActuatorManager.ExecuteHeuristic();
```

Comment:
Discussed offline, but I think the right caller for m_ActuatorManager.ExecuteHeuristic is the HeuristicPolicy itself. 

Actually, as it's written, this might recurse infinitely for Agents that don't override `Heuristic(ActionBuffers)` - HeuristicPolicy calls Agent.Heuristic, Agent.Heuristic calls ActuatorManager.ExecuteHeuristic(), and ActuatorManager.ExecuteHeuristic() calls Agent.Heuristic.

Reply:
Makes sense, I'll update the flow.